lib/widgets/compose_box.dart

```diff
@@ -0,0 +1,281 @@
+    );
+
+    widget.contentController.clear();
+  }
+
+  void _showSendFailedDialog(BuildContext context) {
+    return showErrorDialog(
+        context: context,
+        title: 'Message not sent',
+        message: [
+          ...widget.topicController.validationErrors().map((error) {
+            switch (error) {
+              case TopicValidationError.tooLong:
+                return "Topic length shouldn't be greater than 60 characters.";
+              case TopicValidationError.mandatoryButEmpty:
+                return 'Topics are required in this organization.';
+            }
+          }),
+          ...widget.contentController.validationErrors().map((error) {
+            switch (error) {
+              case ContentValidationError.tooLong:
+              // TODO: UI text copied from web; is it right? Discussion:
+              //   https://chat.zulip.org/#narrow/stream/412-api-documentation/topic/max.20message.20size/near/1505518
+                return "Message length shouldn't be greater than 10000 characters.";
+              case ContentValidationError.empty:
+                return 'You have nothing to send!';
+            }
+          }),
+        ].join('\n\n'));
+  }
+
+  @override
+  Widget build(BuildContext context) {
+    ColorScheme colorScheme = Theme.of(context).colorScheme;
+
+    bool submitButtonDisabled = topicHasValidationErrors || contentHasValidationErrors;
+
+    // Copy FilledButton defaults (_FilledButtonDefaultsM3.backgroundColor)
+    final backgroundColor = submitButtonDisabled
+        ? colorScheme.onSurface.withOpacity(0.12)
+        : colorScheme.primary;
+
+    // Copy FilledButton defaults (_FilledButtonDefaultsM3.foregroundColor)
+    final foregroundColor = submitButtonDisabled
+        ? colorScheme.onSurface.withOpacity(0.38)
+        : colorScheme.onPrimary;
+
+    return Ink(
+      decoration: BoxDecoration(
+        borderRadius: const BorderRadius.all(Radius.circular(8.0)),
+        color: backgroundColor,
+      ),
+      child: IconButton(
+        // Empirically, match the height of the content input. Ideally, this
+        // would be dynamic and respond to the actual height of the content
+        // input. Zeroing the padding lets the constraints take over.
+        constraints: const BoxConstraints(minWidth: 35, minHeight: 35),
+        padding: const EdgeInsets.all(0),
+
+        color: foregroundColor,
+        icon: const Icon(Icons.send),
+        onPressed: () => _handleSendPressed(context),
+      ),
+    );
+  }
+}
+
+/// The compose box for writing a stream message.
+class StreamComposeBox extends StatefulWidget {
+  const StreamComposeBox({super.key});
+
+  @override
+  State<StreamComposeBox> createState() => _StreamComposeBoxState();
+}
+
+class _StreamComposeBoxState extends State<StreamComposeBox> {
+  final _topicController = TopicTextEditingController();
+  final _contentController = ContentTextEditingController();
+
+  @override
+  void dispose() {
```

Comment:
This still needs to listen to the topic text controller too, though — otherwise the content input's placeholder doesn't update.

(Alternatively that could be its own widget.)